subsys/net/l2/ieee802154/ieee802154.c

```diff
@@ -237,6 +238,8 @@ static enum net_verdict ieee802154_recv(struct net_if *iface, struct net_pkt *pk
 
 	ieee802154_acknowledge(iface, &mpdu);
 
+	net_pkt_set_ll_proto_type(pkt, ETH_P_IEEE802154);
+
```

Comment:
I'm not feel comfortable to have a dependency from ethernet inside IEEE 802.15.4.

The set_pkt_ll_addr do this after all:

https://github.com/zephyrproject-rtos/zephyr/blob/2037b20db52a40d8639ea25647b362e4a4fa15e4/subsys/net/l2/ieee802154/ieee802154.c#L100

Is not possible to consult net_linkaddr->addr->type later ?

Reply:
@nandojve I'm afraid there's no short answer to this... I did quite a lot of research before choosing this place as I was equally surprised when first looking at the Linux kernel and socket UAPI. But as a matter of fact `linux/if_ether.h` provides the (socket) PHY layer categorization for many network stacks - not just ethernet (albeit not for all - e.g. bluetooth seems to be an exception). See https://elixir.bootlin.com/linux/v6.0.1/A/ident/ETH_P_IEEE802154 for IEEE 802.15.4 or https://elixir.bootlin.com/linux/v6.0.1/C/ident/ETH_P_CAN for another PHY relevant to Zephyr. These constants even made it into userspace (see e.g. https://github.com/linux-wpan/wpan-tools/blob/master/examples/af_packet_rx.c for IEEE 802.15.4 or more generally `man 7 packet`, `sll_protocol`).

Zephyr has chosen to model their socket API following Linux conventions (which is not a bad choice IMHO). As we provide IEEE 802.15.4 RX socket support in `ieee802154_recv()` in `ieee802154.c`, this is IMO the correct place to set the protocol type as it is consistent with the Linux implementation and Zephyr ethernet stack's `ether_recv()`.

As an aside: The fact that Zephyr's CAN L2 does not set this attribute although it has its own socket stack is IMHO an oversight (as it was prior to this change in IEEE 802.15.4). Bluetooth does not yet support low level sockets in Zephyr, so maybe that's why they don't set a protocol, although Linux's bluetooth implementation does so?

My understanding is that `net_linkaddr.type` is something different. It is not generally true that there is a single protocol for a given link layer address nor the contrary. So `net_linkaddr.type` and `net_pkt.ll_proto_type` are orthogonal IMO.

The doxygen documentation is at best ambiguous and at worst plain wrong calling `net_linkaddr` a "hardware" address rather than a link-layer address. There is no global structure like `net_linkaddr` in Linux AFAICS. The details of LL addresses are usually kept privately in `skbuff->cb` or in opaque "polymorphic" protocol family specific socket addresses which expose the address family of the address so that the address can be cast to the protocol specific implementation. I'm not sure why `net_linkaddr.type` has its own global configuration space rather than referring to AF_.../PF_ constants. Is it maybe just legacy of pre-socket Zephyr code?

I'd therefore argue the other way round: `net_linkaddr.type`/`enum net_link_type` is redundant and should probably be replaced by polymorphic socket addresses whose internal polymorphic structure depends on the publicly visible AF_... constant. These structures already exist in Zephyr and could easily be used to replace `net_linkaddr` AFAICS.

The only uses of `net_linkaddr.type` I can find in the code are to identify 6lo headers and generate IPv6 addresses out of LL addresses. Both would be better served by opaque socket addresses and delegated to L2 rather than kept in net core breaking encapsulation as is now the case. But all this is definitely out-of-scope for the current change set.

Reply:
It seems there is not much space to set that info. I'm thinning that maybe a solution inside **net_pkt** could avoid to contaminate any L2 with Ethernet stuff. Do you think it is valid introduce something like below at **net_pkt**?

```diff
+void net_pkt_set_ll_proto_type_<by_ll_type>(struct net_pkt *pkt, enum net_link_type type) {
+	switch (type) {
+	...
+	case NET_LINK_IEEE802154:
+		net_pkt_set_ll_proto_type(pkt, ETH_P_IEEE802154);
+		break;
+	...
+	}
+}
```

Reply:
IMO this is much worse as it introduces more dependencies on specific L2 stacks into net core while the solution I propose does not break encapsulation anywhere plus it is in line with what ethernet.h obviously claims to be (i.e. a counterpart to if_ether.h). Let's see what others think about this topic. Of course I'll be convinced if your proposal becomes the majority vote.

Reply:
Personally, I'd stick to the current proposal.

@nandojve While the solution you propose may seem simple for IEEE802154, things are getting more complicated for Ethernet, where the actual proto type is not fixed, but rather parsed from the L2 header (EtherType field). And it gets even more complex with VLAN involved. All this parsing logic would need to be covered by this generic function, which is not the best choice IMHO.